### PR TITLE
Order creation - products multi-selection

### PR DESCRIPTION
**Pull request by samiuelson**
### Description

This PR integrates the product selector component into the order creation (and edition) flow.

The goal of this feature is to allow a multi-selection of products while creating an order. 

Previously, the user could select only a single product at once and needed to wait a couple of seconds for the order to get updated each time. Now, the user can select multiple products at once and the order creation flow is significantly faster.

More details in the project thread: pdfdoF-2lX-p2

This PR is based on the following ones, already reviewed:
* https://github.com/woocommerce/woocommerce-android/pull/8545
* https://github.com/woocommerce/woocommerce-android/pull/8553
* https://github.com/woocommerce/woocommerce-android/pull/8519
* https://github.com/woocommerce/woocommerce-android/pull/8443

### Implementation
* The `ProductSelectorFragment` was plugged into the order creation/edition screen and is opened on "+ Add products" button tap.
* Old product picker code was removed
* ProductSelector was improved to allow filtering out draft products and variable products with no variations. This was designed to be configurable, e.g. ProductSelector can still be used in other places and allow selecting draft products. The coupon edition flow was updated as well and doesn't show draft products and variable products without variations.

### Testing instructions

Ensure it's possible to select multiple products and product variations at once while creating new order or editing an existing one (the existing order must not be in a completed state).

### Images/gif

https://user-images.githubusercontent.com/4527432/225295430-1f71e505-77be-48f3-b42a-7b48d997dc7d.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: RELEASE-NOTES.txt [hunk 1]
```diff
@@ -3,6 +3,8 @@
 -----
 - [*][Internal] Do not show full screen input for the payment amount for simple payments flow. [https://github.com/woocommerce/woocommerce-android/pull/8517]
 - [*][Internal] Fix the wording for the cases when we don't have due date on stripe pending requirements state for IPP [https://github.com/woocommerce/woocommerce-android/pull/8516]
+- [**] Added product multi-selection in order creation flow [https://github.com/woocommerce/woocommerce-android/pull/8557]
+- [*] Improved product selector component to allow filtering out redundant products [https://github.com/woocommerce/woocommerce-android/pull/8557]
```
Comment (AnirudhBhat) [inline RELEASE-NOTES.txt:7]: ❓ Can we also add that the local search feature will be available on the product selection screen?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt [hunk 2]
```diff
@@ -350,9 +357,25 @@ class ProductSelectorViewModel @Inject constructor(
         data class ProductVariation(val productId: Long, val variationId: Long) : SelectedItem(variationId)
     }
 
-    private val Set<SelectedItem>.variationIds: List<Long>
-        get() {
-            return filterIsInstance<SelectedItem.ProductOrVariation>().map { it.id } +
-                filterIsInstance<SelectedItem.ProductVariation>().map { it.variationId }
+    @Parcelize
+    sealed class ProductSelectorRestriction : (Product) -> Boolean, Parcelable {
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt:361]: ❤️

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt [hunk 3]
```diff
@@ -76,12 +77,18 @@ class ProductSelectorViewModel @Inject constructor(
     private val loadingState = MutableStateFlow(IDLE)
     private val selectedItems = savedState.getStateFlow(viewModelScope, navArgs.selectedItems.toSet())
     private val filterState = savedState.getStateFlow(viewModelScope, FilterState())
+    private val productsRestrictions = navArgs.restrictions
+    private val products = listHandler.productsFlow.map { products ->
+        products.filter { product ->
```
Comment (AnirudhBhat) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt:82]: ❓ Can we write tests to verify that we filter out only the published products and no variable products with no variation?

**Review by AnirudhBhat — APPROVED** [review]
Thank you for your work on this @samiuelson. I've tested the functionality and it works expected. I've left a couple of minor comments that are non-blocking. 

You might want to look at one of the comments mentioning adding a Local search in the product selection screen to the release notes.

Also, I found this issue when we filter the product, for a few seconds, we display the "no orders" screen and then load the products. This seems to be present in the coupons screen as well.

https://user-images.githubusercontent.com/1331230/225545288-641d99d8-b68e-4459-ad75-09274f691c57.mp4

File: RELEASE-NOTES.txt (hunk 1 above)
Comment (samiuelson) [reply RELEASE-NOTES.txt:7]: Thanks for reminding me about this! Added with 542e53082a964e4b1e8694301f4ae3af54afee41

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt (hunk 3 above)
Comment (samiuelson) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt:82]: Added with 95af3171abc021d735093d9c398ab6bb3caae0ec.

**Comment by samiuelson** [comment]
> Thank you for your work on this @samiuelson. I've tested the functionality and it works expected. I've left a couple of minor comments that are non-blocking.
> 
> You might want to look at one of the comments mentioning adding a Local search in the product selection screen to the release notes.
> 
> Also, I found this issue when we filter the product, for a few seconds, we display the "no orders" screen and then load the products. This seems to be present in the coupons screen as well.

Thank you so much for the code review, @AnirudhBhat  🙌🏼 

I've created an issue for "no orders" empty state you've mentioned and added it to the project [board](https://github.com/woocommerce/woocommerce-android/projects/49). There is also another issue related to `ProductSelector` directly - about UI glitch in the landscape. I've marked both with the new label `feature: product selector`. Let's tackle them in the next iteration.

**Pull request merged**